### PR TITLE
Windows fixes for the cli

### LINKED ISSUE #28: Local Pipeline run failing on Windows
Hi there, I'm attempting to run a pipeline locally on Windows, and am having some trouble. I've pasted the output below and you can find a copy of the static steps [here](https://github.com/spatialos/gdk-for-unity/blob/develop/.buildkite/premerge.steps.yaml).

Any ideas where I'm going wrong? 

Thanks

```
λ bk local run .buildkite/premerge.steps.yaml --debug
2019/02/06 18:29:18 [keyring] Considering backends: [file]
2019/02/06 18:29:18 Serving API on http://127.0.0.1:49256
2019/02/06 18:29:18 Creating bootrap script at C:\Users\JAMIEB~1\AppData\Local\Temp\bootstrap-450177923
>>> Starting local agent 🤖
>>> Starting build 👟
>>> Executing initial command: buildkite-agent pipeline upload ".buildkite/premerge.steps.yaml"

  _           _ _     _ _    _ _                                _
 | |         (_) |   | | |  (_) |                              | |
 | |__  _   _ _| | __| | | ___| |_ ___    __ _  __ _  ___ _ __ | |_
 | '_ \| | | | | |/ _` | |/ / | __/ _ \  / _` |/ _` |/ _ \ '_ \| __|
 | |_) | |_| | | | (_| |   <| | ||  __/ | (_| | (_| |  __/ | | | |_
 |_.__/ \__,_|_|_|\__,_|_|\_\_|\__\___|  \__,_|\__, |\___|_| |_|\__|
                                                __/ |
 http://buildkite.com/agent                    |___/

2019-02-06 18:29:18 NOTICE Starting buildkite-agent v3.8.4 with PID: 40328
2019-02-06 18:29:18 NOTICE The agent source code can be found here: https://github.com/buildkite/agent
2019-02-06 18:29:18 NOTICE For questions and support, email us at: hello@buildkite.com
2019-02-06 18:29:18 INFO   Configuration loaded from: C:\buildkite-agent\buildkite-agent.cfg
2019-02-06 18:29:18 DEBUG  Bootstrap command: C:\Users\JAMIEB~1\AppData\Local\Temp\bootstrap-450177923
2019-02-06 18:29:18 DEBUG  Build path: C:\Users\JAMIEB~1\AppData\Local\Temp\buildkite-build-735988486
2019-02-06 18:29:18 DEBUG  Hooks directory: C:\buildkite-agent\hooks
2019-02-06 18:29:18 DEBUG  Plugins directory: C:\buildkite-agent\plugins
2019-02-06 18:29:18 INFO   Running builds within a pseudoterminal (PTY) has been disabled
2019-02-06 18:29:18 INFO   Registering agent with Buildkite...
2019-02-06 18:29:18 INFO   Started 1 Agent(s)
2019-02-06 18:29:18 INFO   You can press Ctrl-C to stop the agents
2019-02-06 18:29:18 DEBUG  POST http://127.0.0.1:49256/register
2019-02-06 18:29:18 WARN   Post http://127.0.0.1:49256/register: dial tcp 127.0.0.1:49256: socket: The requested service provider could not be loaded or initialized. (Attempt 1/30 Retrying in 10s)
2019-02-06 18:29:28 DEBUG  POST http://127.0.0.1:49256/register
2019-02-06 18:29:28 WARN   Post http://127.0.0.1:49256/register: dial tcp 127.0.0.1:49256: socket: The requested service provider could not be loaded or initialized. (Attempt 2/30 Retrying in 10s)
2019/02/06 18:29:30 [http] GET /register 127.0.0.1:49257
>>> 😓 An unknown agent API endpoint was requested (GET /register).
File an issue at https://github.com/buildkite/cli/issues and we'll see what we can do!
2019/02/06 18:29:30 [http] GET /favicon.ico 127.0.0.1:49257
>>> 😓 An unknown agent API endpoint was requested (GET /favicon.ico).
File an issue at https://github.com/buildkite/cli/issues and we'll see what we can do!
2019-02-06 18:29:38 DEBUG  POST http://127.0.0.1:49256/register
2019-02-06 18:29:38 WARN   Post http://127.0.0.1:49256/register: dial tcp 127.0.0.1:49256: socket: The requested service provider could not be loaded or initialized. (Attempt 3/30 Retrying in 10s)
2019-02-06 18:29:48 DEBUG  POST http://127.0.0.1:49256/register
2019-02-06 18:29:48 WARN   Post http://127.0.0.1:49256/register: dial tcp 127.0.0.1:49256: socket: The requested service provider could not be loaded or initialized. (Attempt 4/30 Retrying in 10s)
2019-02-06 18:29:58 DEBUG  POST http://127.0.0.1:49256/register
2019-02-06 18:29:58 WARN   Post http://127.0.0.1:49256/register: dial tcp 127.0.0.1:49256: socket: The requested service provider could not be loaded or initialized. (Attempt 5/30 Retrying in 10s)
2019-02-06 18:30:08 DEBUG  POST http://127.0.0.1:49256/register
2019-02-06 18:30:08 WARN   Post http://127.0.0.1:49256/register: dial tcp 127.0.0.1:49256: socket: The requested service provider could not be loaded or initialized. (Attempt 6/30 Retrying in 10s)

>>> Gracefully shutting down...

🚨 Initial command failed: context canceled
```

cc @petemounce @DoomGerbil

### PR DESCRIPTION
**Pull request by crufter**
Should fix #28 

Based on @tduffield's very useful comments in #58.

### CODE REVIEW

File: local/run.go [hunk 1]
```diff
@@ -406,11 +408,25 @@ func (a *Agent) Run(ctx context.Context) error {
...
+			"SYSTEMDRIVE="+os.Getenv("SYSTEMDRIVE"),
+		)
+	}
+
 	// this function is called at the end of Run()
 	// it kills the agent
 	a.stopFunc = func() error {
 		defer os.Remove(bootstrap.Name())
-		_ = cmd.Process.Signal(os.Interrupt)
+		_ = cmd.Process.Kill()
```
Comment (crufter) [inline local/run.go]: Signals do not really work on windows, I have ran into this issue many times in the past.

File: local/run.go [hunk 2]
```diff
@@ -442,17 +458,29 @@ func (a *Agent) Stop() error {
 }
 
 func createAgentBootstrap(checkoutPath string) (*os.File, error) {
-	tmpFile, err := ioutil.TempFile(os.TempDir(), "bootstrap-")
+	tempFileNamePattern := "bootstrap-"
+	if runtime.GOOS == "windows" {
+		tempFileNamePattern = "bootstrap-*.bat"
```
Comment (crufter) [inline local/run.go]: The `.bat` extension is really important, without it the agent won't find the boostrap script.

Comment (crufter) [inline local/run.go]: The `.bat` extension is really important, without it the agent won't find the boostrap script.

File: local/run.go [hunk 3]
```diff
@@ -406,11 +408,25 @@ func (a *Agent) Run(ctx context.Context) error {
...
+	if runtime.GOOS == "windows" {
+		cmd.Env = append(cmd.Env,
+			"PATH="+os.Getenv("PATH"),
+			"SystemRoot="+os.Getenv("SystemRoot"),
+			"WINDIR="+os.Getenv("WINDIR"),
+			"COMSPEC="+os.Getenv("COMSPEC"),
+			"PATHEXT="+os.Getenv("PATHEXT"),
+			"TMP="+os.Getenv("TMP"),
+			"TEMP="+os.Getenv("TEMP"),
+			"SYSTEMDRIVE="+os.Getenv("SYSTEMDRIVE"),
```
Comment (crufter) [inline local/run.go]: Not sure if this is needed tbh, will test.

File: local/run.go (hunk 2 above)
Comment (petemounce) [reply local/run.go]: `.cmd` is the more modern choice (unless powershell is an option!): https://serverfault.com/questions/17899/whats-the-difference-between-cmd-and-bat-files

File: local/run.go (hunk 1 above)
Comment (petemounce) [reply local/run.go]: Worth wrapping inside a runtime.GOOS=="windows" conditional so the other OS' are not brutalised for Windows' failing?

Comment (crufter) [reply local/run.go]: Good idea! Will do.

File: local/run.go (hunk 2 above)
Comment (crufter) [reply local/run.go]: Might be an option, have to check the agent source code.

**Comment by crufter** [comment]
Unfortunately I do not have access to a Windows machine anymore, so I can't make any more improvements currently.

I have tested all these changes but the last commit.

**Comment by lox** [comment]
This looks great, we will merge and test

**Pull request merged**